### PR TITLE
(feat): Support for scipy.sparse.{csr,csc}_array

### LINKED ISSUE #746: Support for scipy sparse arrays
Apparently scipy now has [sparse arrays with array indexing](https://docs.scipy.org/doc/scipy/reference/sparse.html)?

Unfortunately they do seem to inherit from the matrix classes, which makes checking their types a bit fraught.

### PR DESCRIPTION
**Pull request by ivirshup**
Starting support for scipy.sparse csc_array and csr_array. These are the new array-like interfaces for sparse matrices from scipy.

Fixes #746, #859

Still very WIP

TODO:

 - [x] Figure out compat with scipy<1.11
 - [ ] Demo of how to load an array from disk
 - [x] Support for `.X`
 - [x] Tests (figure out which ones need adding)
 - [x] Fix concatenation tests
 - [x] Fix view tests
 - [x] Test for indices dtype after concatenation (e.g. #859)
 - [x] Release note
 - [x] Remove support for `dask.array[sparse.sparray]`

### CODE REVIEW

File: anndata/tests/helpers.py [hunk 1]
```diff
@@ -215,35 +240,60 @@ def gen_adata(
...
+                sparse.random(N, 100, format=sparse_fmt, random_state=random_state)
+            )
     layers = dict(
         array=np.random.random((M, N)),
-        sparse=sparse.random(M, N, format=sparse_fmt),
+        sparse=sparse.random(M, N, format=sparse_fmt, random_state=random_state),
         da=da.random.random((M, N)),
     )
+    if CAN_USE_SPARSE_ARRAY:
+        if sparse.csr_array in layers_types or sparse.csr_matrix in layers_types:
```
Comment (ilan-gold) [inline anndata/tests/helpers.py:276]: Should we add if only `csr_matrix` is present?

**Comment by ilan-gold** [comment]
Ok @ivirshup I think this is read for a review.  This PR should only focus on the use of the object in the `AnnData` class and not on reading.  That will come next, but here there is writing functionality (which made sense to me to put here)..

**Comment by ilan-gold** [comment]
The test failures are from "FAILED anndata/tests/test_backed_sparse.py::test_append_overflow_check[csr_array-zarr] - numpy.core._exceptions._ArrayMemoryError: Unable to allocate 16.0 GiB for a..." - we could xfail these as well, but it seems like since this works for matrix, maybe there's a bug here?

File: anndata/tests/test_x.py [hunk 2]
```diff
@@ -30,11 +32,16 @@ def diskfmt(request):
...
+def test_setter_singular_dim(shape, orig_array_type, new_array_type, request):
     # https://github.com/scverse/anndata/issues/500
     adata = gen_adata(shape, X_type=orig_array_type)
     adata.X = new_array_type(np.ones(shape))
     np.testing.assert_equal(asarray(adata.X), 1)
+    id = request.node.callspec.id.split("-")[0]
+    if id == "ndarray":
+        assert isinstance(adata.X, np.ndarray)
+    else:
+        assert isinstance(adata.X, new_array_type)
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/tests/test_x.py:44]: `X` is cast immediately after that to `new_array_type`, which is what the issue that this links to says we should be loooking for, no?

File: anndata/_core/sparse_dataset.py [hunk 3]
```diff
@@ -225,9 +225,27 @@ def _get_sliceXarray(self, row: slice, col: Sequence[int]) -> ss.csc_matrix:
...
+class backed_csc_matrix(backed_csc, ss.csc_matrix):
+    """backed_csc_matrix"""
+
+
+class backed_csr_array(backed_csr, ss.csr_array):
+    """backed_csr_array"""
+
+
+class backed_csc_array(backed_csc, ss.csc_array):
+    """backed_csc_array"""
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/_core/sparse_dataset.py:241]: I agree but I think that is out-of-scope here.  I can open a separate PR for that.  I think there will be a lot of edge cases around normalizing the indexers, which we currently get for free with these classes (although we could call `_validate_indices` from the matrix class to solve that).

File: anndata/tests/helpers.py [hunk 4]
```diff
@@ -153,26 +175,26 @@ def gen_typed_df_t2_size(m, n, index=None, columns=None) -> pd.DataFrame:
     return df
 
 
+default_key_types = (
+    sparse.csr_matrix,
+    np.ndarray,
+    pd.DataFrame,
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/tests/helpers.py:181]: This is used for `layers` as well so that won't work.  Could refactor but I think what we have [here](https://github.com/scverse/anndata/pull/1028/files/83c6363a4767efbbbfbb110fc89e23e0549c3442#diff-1913caafc2716cc3f96cb7cf500214abb2ff07656987875eb03222f5e8b6e6c5R194-R196) is fine

File: anndata/_core/sparse_dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/_core/sparse_dataset.py:241]: See #1456.  We'll put it in for the next sprint.

File: anndata/tests/helpers.py [hunk 5]
```diff
@@ -215,35 +240,60 @@ def gen_adata(
...
         array=np.random.random((M, 50)),
-        sparse=sparse.random(M, 100, format=sparse_fmt),
+        sparse=sparse.random(M, 100, format=sparse_fmt, random_state=random_state),
         df=gen_typed_df(M, obs_names),
         awk_2d_ragged=gen_awkward((M, None)),
         da=da.random.random((M, 50)),
     )
     obsm = {k: v for k, v in obsm.items() if type(v) in obsm_types}
+    if CAN_USE_SPARSE_ARRAY:
+        if sparse.csr_array in obsm_types or sparse.csr_matrix in obsm_types:
```
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/tests/helpers.py:253]: Yes!

File: anndata/_core/sparse_dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/_core/sparse_dataset.py:241]: Which indexing cases are different here?

**Comment by flying-sheep** [comment]
sorry for the force push, VS code messed things up and I simply undid the mess to get a clean merge commit

**Comment by ilan-gold** [comment]
No worries @flying-sheep - much appreciated!

File: anndata/_core/sparse_dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/_core/sparse_dataset.py]: Nothing changes here but in the spirit of keeping this PR smaller, I figured we'd make an issue and hit it next sprint.  We can do a little drive-by, but I don't think it'll be so little.  That was why I split the reading-to-sparse-arrays feature separate

**Comment by ilan-gold** [comment]
> Demo of how to load an array from disk

Probably belongs in the follow up PR given that we did not implement that functionality here

File: anndata/tests/test_x.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivirshup) [reply anndata/tests/test_x.py]: I meant for `id`

[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/tests/test_x.py]: ok, we can do that.

File: src/anndata/tests/test_concatenate.py [hunk 6]
```diff
@@ -1452,12 +1465,13 @@ def test_concat_null_X():
...
-def test_concat_X_dtype(array_type, sparse_indexer_type):
+def test_concat_X_dtype(cpu_array_type, sparse_indexer_type):
     adatas_orig = {
-        k: AnnData(array_type(np.ones((20, 10), dtype=np.int8))) for k in list("abc")
+        k: AnnData(cpu_array_type(np.ones((20, 10), dtype=np.int8)))
+        for k in list("abc")
     }
     for adata in adatas_orig.values():
-        adata.raw = AnnData(array_type(np.ones((20, 30), dtype=np.float64)))
+        adata.raw = AnnData(cpu_array_type(np.ones((20, 30), dtype=np.float64)))
```
Comment (ilan-gold) [inline src/anndata/tests/test_concatenate.py]: Can we just change the dtype from `uint8` here?  Or is there a reason to keep this?  I think that this is what is causing GPU test failures

File: src/anndata/tests/test_concatenate.py [hunk 7]
```diff
@@ -1452,12 +1465,13 @@ def test_concat_null_X():
 
 # https://github.com/scverse/ehrapy/issues/151#issuecomment-1016753744
 @pytest.mark.parametrize("sparse_indexer_type", [np.int64, np.int32])
-def test_concat_X_dtype(array_type, sparse_indexer_type):
+def test_concat_X_dtype(cpu_array_type, sparse_indexer_type):
     adatas_orig = {
-        k: AnnData(array_type(np.ones((20, 10), dtype=np.int8))) for k in list("abc")
+        k: AnnData(cpu_array_type(np.ones((20, 10), dtype=np.int8)))
```
Comment (ilan-gold) [inline src/anndata/tests/test_concatenate.py:1470]: Can we just change the dtype from uint8 here? Or is there a reason to keep this? I think that this is what is causing GPU test failures

File: anndata/tests/test_x.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ilan-gold) [reply anndata/tests/test_x.py]: Nevermind, it's better to use the `id` because some of these `new_array_type` are functions, which have `__name__` and others are not since they come from `partial`.

File: src/anndata/_core/anndata.py [hunk 8]
```diff
@@ -1773,8 +1778,11 @@ def concatenate(
 
         # Backwards compat (some of this could be more efficient)
         # obs used to always be an outer join
+        sprase_class = sparse.csr_matrix
+        if any(isinstance(a.X, SpArray) for a in all_adatas):
```
Comment (ilan-gold) [inline src/anndata/_core/anndata.py:1782]: Should we remove this for codecov to be happy since the function is already deprecated?

Comment (ilan-gold) [reply src/anndata/_core/anndata.py:1782]: I don't think this was tested for anyway

File: src/anndata/_core/sparse_dataset.py [hunk 9]
```diff
@@ -302,24 +310,23 @@ def mean_slice_length(slices):
...
-    raise ValueError(f"Data type {type(data)} is not supported.")
-
-
-def get_memory_class(format: str) -> type[ss.spmatrix]:
+def get_memory_class(format: str, use_sparray_in_io=False) -> type[ss.spmatrix]:
     for fmt, _, memory_class in FORMATS:
         if format == fmt:
-            return memory_class
+            if use_sparray_in_io and issubclass(memory_class, SpArray):
+                return memory_class
```
Comment (ilan-gold) [inline src/anndata/_core/sparse_dataset.py:317]: Expected codecov miss and will be resolved by the next PR

File: src/anndata/_core/sparse_dataset.py [hunk 10]
```diff
@@ -302,24 +310,23 @@ def mean_slice_length(slices):
...
     raise ValueError(f"Format string {format} is not supported.")
 
 
-def get_backed_class(format: str) -> type[BackedSparseMatrix]:
+def get_backed_class(format: str, use_sparray_in_io=False) -> type[BackedSparseMatrix]:
     for fmt, backed_class, _ in FORMATS:
         if format == fmt:
-            return backed_class
+            if use_sparray_in_io and issubclass(backed_class, SpArray):
+                return backed_class
```
Comment (ilan-gold) [inline src/anndata/_core/sparse_dataset.py:327]: Expected codecov miss and will be resolved by the next PR

File: src/anndata/compat/__init__.py [hunk 11]
```diff
@@ -15,18 +15,35 @@
...
+
+if not CAN_USE_SPARSE_ARRAY:
+
+    class SpArray:
+        @staticmethod
+        def __repr__():
+            return "mock scipy.sparse.sparray"
+else:
+    SpArray = scipy.sparse.sparray
+
```
Comment (ilan-gold) [inline src/anndata/compat/__init__.py:39]: Well, the lower-scipy-version miss makes sense if the repr is not called, but not sure what's up with the second half

Comment (ilan-gold) [reply src/anndata/compat/__init__.py:39]: In reference to codecov

File: src/anndata/tests/helpers.py [hunk 12]
```diff
@@ -628,18 +714,66 @@ def as_sparse_dask_array(a) -> DaskArray:
...
     return a.map_blocks(sparse.csr_matrix)
 
 
+@as_sparse_array_dask_array.register(DaskArray)
+def _(a):
+    return a.map_blocks(sparse.csr_array)
+
+
+@contextmanager
+def pytest_8_raises(exc_cls, *, match: str | re.Pattern = None):
```
Comment (ilan-gold) [inline src/anndata/tests/helpers.py:763]: remove?

Comment (ilan-gold) [reply src/anndata/tests/helpers.py:763]: :(
